dos/src/main.rs

```diff
@@ -23,14 +33,128 @@ fn get_repair_contact(nodes: &[ContactInfo]) -> ContactInfo {
     contact
 }
 
+/// Options for data_type=transaction
+#[derive(Serialize, Deserialize, Debug)]
+struct TransactionParams {
+    unique_transactions: bool, // use unique transactions
+    num_sign: usize,           // number of signatures in a transaction
+    valid_block_hash: bool,    // use valid blockhash or random
+    valid_signatures: bool,    // use valid signatures or not
+    with_payer: bool,          // provide a valid payer
+}
+
+struct TransactionGenerator {
+    blockhash: Hash,
+    last_generated: Instant,
+    transaction_params: TransactionParams,
+    cached_transaction: Option<Transaction>,
+}
+
+impl TransactionGenerator {
+    fn new(transaction_params: TransactionParams) -> Self {
+        TransactionGenerator {
+            blockhash: Hash::default(),
+            last_generated: (Instant::now() - Duration::from_secs(100)),
+            transaction_params,
+            cached_transaction: None,
+        }
+    }
+
+    fn generate(&mut self, payer: &Keypair, rpc_client: &Option<RpcClient>) -> Transaction {
+        if !self.transaction_params.unique_transactions && self.cached_transaction != None {
+            return self.cached_transaction.as_ref().unwrap().clone();
+        }
+
+        // generate a new blockhash every 1sec
+        if self.transaction_params.valid_block_hash
+            && self.last_generated.elapsed().as_millis() > 1000
+        {
+            self.blockhash = rpc_client.as_ref().unwrap().get_latest_blockhash().unwrap();
+            self.last_generated = Instant::now();
+        }
+
+        let lamports = 5;
+        let transfer_instruction = SystemInstruction::Transfer { lamports };
+        let program_ids = vec![system_program::id(), stake::program::id()];
+
+        // transaction with payer, in this case signatures are valid and num_sign is irrelevant
+        // random payer will cause error "attempt to debit an account but found not record of a prior credit"
+        // if payer is correct, it will trigger error with not enough signatures
+        let transaction = if self.transaction_params.with_payer {
+            let instruction = Instruction::new_with_bincode(
+                program_ids[0],
+                &transfer_instruction,
+                vec![
+                    AccountMeta::new(program_ids[0], false),
+                    AccountMeta::new(program_ids[1], false),
+                ],
+            );
+            Transaction::new_signed_with_payer(
```

Comment:
I think this case may be limited by signing speed. We may need to push this to a separate thread, as in other bench/dos clients.

Reply:
I would propose to postpone it for the next PR since this one is already heavy